### PR TITLE
[Zoe] Allow sparse keywords

### LINKED ISSUE #393: [Zoe] Allow smart contracts to reallocate units for a subset of the canonical list of assays
(It's not clear whether this should be a helper filling out the unitMatrix or whether it should be built in.)

### LINKED ISSUE #785: [Zoe] IssuerKeywordRecord should be an optional argument in `makeInstance`
The multipool autoswap shouldn't have to start with any issuers. The agent who makes the instance should use the external method to add a new liquidity pool. Currently, Zoe requires an IssuerKeywordRecord to make an instance, but it should not.

### LINKED ISSUE #800: [Zoe] Sparse Keywords Spec
Zoe currently has the concept of `keywords` which are used as indexes in place of the array indexes that we had previously. The keywords are used to match `proposal` elements, `payments`, `amounts` in the offerRecord within Zoe, and `payouts` to the user. Keywords are per contract, and are currently objective: they are the same for all users of a contract. 

We expect that future contracts (specifically multipool Autoswap https://github.com/Agoric/agoric-sdk/issues/391) will have many keywords, potentially hundreds. Therefore, we want to ensure that the `keywordRecords` (records using keywords as keys) throughout Zoe are sparse, much sparser than all the keywords for a contract. 

The question is where a subset of these keywords should be used, and how this subset is derived at different points.

Below are key points in Zoe where keywords are used:
#### User
1. The proposal sent in by the user
2. The offer payments sent in by the user
3. The payouts to the user

#### Zoe
1. `offer.proposal` (the user's offer proposal)
2. `offer.amounts` (the current allocation)

#### Contract
1. `offer.proposal` (the user's offer proposal, possibly altered in the query)
2. `offer.amounts` (the current allocation, possibly altered in the query)
3. `newAmountKeywordRecord` arguments to `zcf.reallocate`

The current Zoe makes this easy by filling all the keywords that the contract has at the time that the user escrows with Zoe, so at no point after that step do any partial keywords exist. However, that strategy will not work for contracts with hundreds of keywords. 

We decided March 27th that the solution, the "sparse keywords" PR, should include the following:

- [x]  All of the keywords for the contract should be called `allKeywords`*
- [x] Any particular subset is `sparseKeywords`*
- [x] `offerRecord` will still include `proposal` and `amounts`. However, the Zoe Contract Facet (zcf) API will change.
- [x] `zcf.getOffers` and `zcf.getOffer` will only return `proposal`, `handle`, and `instanceHandle`, not `amounts`
- [x] Two new methods will be added: `zcf.getAmounts` and `zcf.getAmountsForOffer`, which will take as arguments `offerHandles`, and `sparseKeywords`. The representation of amounts that is returned should have filled in empty values for all the `sparseKeywords` and anything not listed in `sparseKeywords` is dropped. Note that nothing changes in Zoe's offerRecord table. This is merely changing what the contract gets as a response to the query. 
- [x] The user can use sparse keywords (whichever they want) in their proposal. Zoe will not fill these in, but will store the proposal as is, after cleaning it. 
- [x] The user's offer payments must match the user's `proposal.give`. Zoe will check that these match, but will not do any additional filling. 
- [x] When Zoe creates `offerRecord.amounts` after depositing the user's payments, Zoe will fill in the `wants` of the user's offer with empty. Zoe will not fill in all keywords, just the ones mentioned by the user.
- [x] `zcf.reallocate` will have a similar signature to `zcf.getAmounts`: `offerHandles`, `sparseKeywords` and `newAmountKeywordRecords` will be the arguments. `zcf.reallocate` will throw an error if any of the `newAmountKeywordRecords` do not have a value for all the keywords in `sparseKeywords`. An error will also be thrown if any `newAmountKeywordRecords` have keywords that are not in `sparseKeywords`.  We expect that contracts will be using `zcf.getAmounts` and then calling `zcf.reallocate`, so these requirements should not be burdensome if the same `sparseKeywords` is used in both calls, since `zcf.getAmounts` fills in the values appropriately.
- [x] The reallocation done in `zcf.reallocate` will change the keywords in `offerRecord.amounts`
- [x] The keywords in the payouts to users will be the same as the current `offerRecord.amounts` for that user, because that is what our offer safety and rights conservation checks were made over.
- [x] Note that the keywords in the payouts might be different than the keywords original mentioned by the user in their offer

Asterisk: Kate's late addition

### LINKED ISSUE #849: [Zoe] `zcf.getAmountMaths` should take an array of keywords as an argument, not issuerKeywordRecord

### PR DESCRIPTION
**Pull request by katelynsills**
In reviewing, let's make sure this implementation matches the "spec" outlined in this issue: https://github.com/Agoric/agoric-sdk/issues/800

Closes https://github.com/Agoric/agoric-sdk/issues/800
Closes #393 
Closes #785 
Closes https://github.com/Agoric/agoric-sdk/issues/849

Closes partially #847

### CODE REVIEW

File: packages/zoe/src/contracts/autoswap.js [hunk 1]
```diff
@@ -64,7 +64,7 @@ export const makeContract = harden(zoe => {
             } else {
               return rejectOffer(inviteHandle);
             }
-            if (!amountMaths.Liquidity.isEmpty(proposal.want.Liquidity)) {
+            if (proposal.want.Liquidity !== undefined) {
```
Comment (Chris-Hibbert) [inline packages/zoe/src/contracts/autoswap.js:68]: Isn't this equivalent to 
```
if (proposal.want.Liquidity) {
```

File: packages/zoe/src/objArrayConversion.js [hunk 2]
```diff
@@ -18,3 +18,50 @@ export const objToArray = (obj, keywords) => {
...
+
+export const assertSubset = (whole, part) => {
+  part.forEach(key => {
+    assert.typeof(key, 'string');
+    assert(whole.includes(key), details`key ${key} was not an expected key`);
+  });
+};
+
+// Keywords must equal the keys of obj
+export const filterObjToArray = (obj, keywords) => {
```
Comment (Chris-Hibbert) [inline packages/zoe/src/objArrayConversion.js]: `filter...` seems like the wrong verb when no elements will be removed. This ensures that all the keys are the same, so nothing will be lost.

File: packages/zoe/src/objArrayConversion.js [hunk 3]
```diff
@@ -18,3 +18,50 @@ export const objToArray = (obj, keywords) => {
...
+  return keywords.map(keyword => {
+    assert(
+      obj[keyword] !== undefined,
+      details`obj[keyword] must be defined for keyword ${keyword}`,
+    );
+    return obj[keyword];
+  });
+};
+
+// return a new object with only the keys in subsetKeywords
```
Comment (Chris-Hibbert) [inline packages/zoe/src/objArrayConversion.js]: Should it enforce that all the subsetKeywords be present in obj?  If not, it's worth saying so in the comment.

File: packages/zoe/src/cleanProposal.js [hunk 4]
```diff
@@ -15,76 +15,98 @@ import { arrayToObj } from './objArrayConversion';
 // `{ waived: null }` `{ onDemand: null }` `{ afterDeadline: { timer
 // :Timer, deadline :Number } }
 
-// Assert that the keys of record, if present, are in expectedKeys.
-// Return the keys after asserting this.
-const checkKeys = (expectedKeys, record) => {
-  // Assert that keys, if present, match expectedKeys.
+export const assertCapASCII = keyword => {
```
Comment (Chris-Hibbert) [inline packages/zoe/src/cleanProposal.js:7]: cleanProposal.js is exporting a few things that aren't covered by the current name. Would it be better to rename it, or to split its functions up further?

The comment above no longer covers the file, so it should be moved closer to the function it describes. (Or the fn moved up here if it doesn't depend on the other fns.)

File: packages/zoe/src/zoe.js [hunk 5]
```diff
@@ -227,22 +226,20 @@ const makeZoe = (additionalEndowments = {}) => {
...
       addNewIssuer: (issuerP, keyword) =>
         issuerTable.getPromiseForIssuerRecord(issuerP).then(issuerRecord => {
-          assertKeyword(keyword);
-          const { issuerKeywordRecord, keywords } = instanceTable.get(
-            instanceHandle,
+          assertCapASCII(keyword);
+          const { issuerKeywordRecord } = instanceTable.get(instanceHandle);
+          assert(
+            !getKeywords(issuerKeywordRecord).includes(keyword),
+            details`keyword must be unique`,
```
Comment (Chris-Hibbert) [inline packages/zoe/src/zoe.js]: include the duplicated keyword in the message

File: packages/zoe/test/swingsetTests/zoe/vat-alice.js [hunk 6]
```diff
@@ -403,6 +403,7 @@ const build = async (E, log, zoe, issuers, payments, installations, timer) => {
     // remove the liquidity
     const aliceRemoveLiquidityProposal = harden({
       give: { Liquidity: liquidity(10) },
+      want: { TokenA: moola(0), TokenB: simoleans(0) },
```
Comment (Chris-Hibbert) [inline packages/zoe/test/swingsetTests/zoe/vat-alice.js:406]: I thought the point of the change was to make this less necessary. If this line is optional, but you're including it for pedagogical reasons, or to test that presence is allowed, please add a comment to the test.

**Review by Chris-Hibbert — CHANGES_REQUESTED** [review]

File: packages/zoe/test/swingsetTests/zoe/vat-alice.js (hunk 6 above)
Comment (katelynsills) [reply packages/zoe/test/swingsetTests/zoe/vat-alice.js:406]: I think there's a misunderstanding here. In the current Zoe on master, all keywords will be filled in in the user's proposal before the contract sees it. In this PR, we purposefully don't fill in the user's proposal, so actually, it is more necessary for the user to use the keywords expected by the contract. In this case, the Autoswap contract expects that all three keywords are used when removing liquidity, so Alice must specify TokenA and TokenB, whereas before she didn't. If Autoswap didn't expect all three keywords are used, there would be no need to add this line. 

To put it another way, previously you had to use every keyword in the contract in your proposal, then we had a PR that allowed you to use none of the keywords in your proposal because they would all be filled in for you, and now we are allowing the users to use whatever subset of keywords they want, but we are holding them to that usage and not filling anything in for them in the proposal.

File: packages/zoe/src/contracts/autoswap.js (hunk 1 above)
Comment (katelynsills) [reply packages/zoe/src/contracts/autoswap.js:68]: I think that would test truthy or falsiness, which would be less precise. I want to ensure it's undefined, but `if (proposal.want.Liquidity) {` would not throw if proposal.wantLiquidity is anything falsy: false, 0, -0, 0n, "", null, undefined, and NaN.

File: packages/zoe/src/objArrayConversion.js (hunk 2 above)
Comment (katelynsills) [reply packages/zoe/src/objArrayConversion.js]: Good point

File: packages/zoe/src/cleanProposal.js (hunk 4 above)
Comment (katelynsills) [reply packages/zoe/src/cleanProposal.js:7]: Yeah, I'm not sure how to split up the content of this file versus `objArrayConversion.js`. They are both more generalized helpers now. 

I will move that comment!

File: packages/zoe/src/objArrayConversion.js (hunk 3 above)
Comment (katelynsills) [reply packages/zoe/src/objArrayConversion.js]: Yes, it should.

File: packages/zoe/test/swingsetTests/zoe/vat-alice.js (hunk 6 above)
Comment (Chris-Hibbert) [reply packages/zoe/test/swingsetTests/zoe/vat-alice.js:406]: Got it. Thanks for explaining. We'll have to make sure to be clear about that, though our users won't have gone through the various transitions, so they only need to understand the end state.

**Review by Chris-Hibbert — APPROVED** [review]

**Review by erights — APPROVED** [review]
LGTM
